dev/PowerNotifications/PowerNotifications.h

```diff
@@ -206,42 +237,80 @@ namespace winrt::Microsoft::Windows::System::Power
 
             event_token AddCallback(PowerFunctionDetails fn, const PowerEventHandler& handler)
             {
-                auto& eventObj{ fn.event() };
-                std::scoped_lock<std::mutex> lock(m_mutex);
-                if (!RegisteredForEvents(eventObj))
+                try
                 {
-                    fn.registerListener();
+                    PowerNotificationsTelemetry::AddCallbackTrace(fn.name.c_str());
+                    auto& eventObj{ fn.event() };
+                    std::scoped_lock<std::mutex> lock(m_mutex);
+                    if (!RegisteredForEvents(eventObj))
+                    {
+                        PowerNotificationsTelemetry::RegisterTrace(fn.name.c_str());
+                        fn.registerListener();
+                    }
+                    return eventObj.add(handler);
+                }
+                catch (std::exception& ex)
+                {
+                    PowerNotificationsTelemetry::FailureTrace(fn.name.c_str(), L"AddCallback", ex.what());
+                    throw ex;
                 }
-                return eventObj.add(handler);
             }
 
             void RemoveCallback(PowerFunctionDetails fn, const event_token& token)
             {
-                auto& eventObj{ fn.event() };
-                std::scoped_lock<std::mutex> lock(m_mutex);
-                eventObj.remove(token);
-                if (RegisteredForEvents(eventObj))
+                try
+                {
+                    PowerNotificationsTelemetry::RemoveCallbackTrace(fn.name.c_str());
+                    auto& eventObj{ fn.event() };
+                    std::scoped_lock<std::mutex> lock(m_mutex);
+                    eventObj.remove(token);
+                    if (RegisteredForEvents(eventObj))
+                    {
+                        PowerNotificationsTelemetry::UnregisterTrace(fn.name.c_str());
+                        fn.unregisterListener();
+                    }
+                }
+                catch (std::exception& ex)
                 {
-                    fn.unregisterListener();
+                    PowerNotificationsTelemetry::FailureTrace(fn.name.c_str(), L"RemoveCallback", ex.what());
+                    throw ex;
                 }
             }
 
             void RaiseEvent(PowerFunctionDetails fn)
             {
-                std::thread thread([fn]() {
-                    fn.event()(nullptr, nullptr);
```

Comment:
https://github.com/microsoft/cppwinrt/pull/858#discussion_r583120467 - there's a way to do this using C++/winrt and coroutines, so your methods become:

```c++
winrt::fire_and_forget RaiseEvent(PowerFunctionDetails fn)
{
    auto lifetime = get_strong();
    co_await winrt::resume_background();
    fn.event()(nullptr, nullptr);
}
```

Reply:
Just curious: is this necessarily better?